File: core/agent-core-python/agent_core/tools.py
from __future__ import annotations

import json
from datetime import UTC, datetime
from typing import Any

from langchain_core.tools import BaseTool, StructuredTool
from pydantic import BaseModel, Field

from .mcp import load_mcp_tools_from_env
from .types import McpToolPlugin, ToolBuildContext, ToolRegistry


class _EchoInput(BaseModel):
    text: str = Field(..., description="Text to echo back.")


class _CalcInput(BaseModel):
    expression: str = Field(..., description="Arithmetic expression using + - * / and parentheses.")


class _RememberFactInput(BaseModel):
    content: str = Field(..., description="Fact to remember for this thread.")
    category: str = Field(default="context", description="Memory fact category.")
    confidence: float = Field(default=0.7, ge=0.0, le=1.0, description="Confidence score for this memory.")


class _ReadSkillInput(BaseModel):
    name: str = Field(..., description="Skill name to read.")


def _to_string(value: Any) -> str:
    if isinstance(value, str):
        return value
    return json.dumps(value, ensure_ascii=False, indent=2)


class DefaultAgentToolRegistry(ToolRegistry):
    def __init__(self) -> None:
        self._plugins: list[McpToolPlugin] = []

    def use_mcp_plugin(self, plugin: McpToolPlugin) -> "DefaultAgentToolRegistry":
        self._plugins.append(plugin)
        return self

    async def build_mcp_tools(self) -> list[tuple[str, BaseTool]]:
        tools: dict[str, tuple[str, BaseTool]] = {}

        for plugin in self._plugins:
            for tool in await plugin.load_tools():
                tools.setdefault(tool.name, (plugin.name, tool))

        for tool in await load_mcp_tools_from_env():
            tools.setdefault(tool.name, ("mcp-server", tool))

        return list(tools.values())

    async def build_tools(self, context: ToolBuildContext) -> list[BaseTool]:
        tools: dict[str, BaseTool] = {}

        async def get_time() -> str:
            return datetime.now(UTC).isoformat()

        async def echo_text(text: str) -> str:
            return text

        async def calculate(expression: str) -> str:
            if not all(char in "0123456789+-*/(). " for char in expression):
                raise ValueError("Expression contains unsupported characters")
            result = eval(expression, {"__builtins__": {}}, {})
            if not isinstance(result, int | float):
                raise ValueError("Expression result is not numeric")
            return str(result)

        async def remember_fact(content: str, category: str = "context", confidence: float = 0.7) -> str:
            if context.memory_store is None:
                return "Memory store is not configured."
            fact = await context.memory_store.create_fact(
                context.thread_id,
                content=content,
                category=category,
                confidence=confidence,
                metadata=context.metadata,
            )
            return _to_string(
                {
                    "id": fact.id,
                    "thread_id": fact.thread_id,
                    "content": fact.content,
                    "category": fact.category,
                    "confidence": fact.confidence,
                }
            )

        async def list_memory() -> str:
            if context.memory_store is None:
                return "Memory store is not configured."
            facts = await context.memory_store.list_facts(context.thread_id, limit=20)
            return _to_string([fact.__dict__ for fact in facts])

        async def list_skills() -> str:
            if context.skill_registry is None:
                return "Skill registry is not configured."
            skills = context.skill_registry.list_skills(
                enabled_only=bool(context.enabled_skills),
                enabled_names=context.enabled_skills,
            )
            return _to_string(
                [
                    {"name": skill.name, "description": skill.description, "path": str(skill.path)}
                    for skill in skills
                ]
            )

        async def read_skill(name: str) -> str:
            if context.skill_registry is None:
                return "Skill registry is not configured."
            skill = context.skill_registry.get_skill(name)
            if skill is None:
                raise ValueError(f"Skill not found: {name}")
            return skill.content

        builtins = [
            StructuredTool.from_function(
                coroutine=get_time,
                name="get_time",
                description="Get the current UTC datetime in ISO-8601 format.",
            ),
            StructuredTool.from_function(
                coroutine=echo_text,
                name="echo_text",
                description="Echo input text for tool-calling validation.",
                args_schema=_EchoInput,
            ),
            StructuredTool.from_function(
                coroutine=calculate,
                name="calculate",
                description="Evaluate a basic arithmetic expression.",
                args_schema=_CalcInput,
            ),
            StructuredTool.from_function(
                coroutine=remember_fact,
                name="remember_fact",
                description="Persist a fact into thread memory for later turns.",
                args_schema=_RememberFactInput,
            ),
            StructuredTool.from_function(
                coroutine=list_memory,
                name="list_memory",
                description="List the saved memory facts for the current thread.",
            ),
            StructuredTool.from_function(
                coroutine=list_skills,
                name="list_skills",
                description="List the available skills and their summaries.",
            ),
            StructuredTool.from_function(
                coroutine=read_skill,
                name="read_skill",
                description="Read the full content of a skill by name.",
                args_schema=_ReadSkillInput,
            ),
        ]

        for tool in builtins:
            tools[tool.name] = tool

        for _, tool in await self.build_mcp_tools():
            tools.setdefault(tool.name, tool)

        return list(tools.values())

File: core/agent-core-ts/ts/tools.ts
import { tool, type StructuredToolInterface } from "@langchain/core/tools";
import { z } from "zod";
import type {
  AgentToolRegistry,
  BuildToolOptions,
  LocalToolSpec,
  McpPluginInfo,
  McpToolPlugin,
  McpToolPluginLoadContext,
  ToolInvocationContext
} from "./types.js";
import { ToolExecutionCoordinator, wrapToolWithPolicy } from "./utils/tool-execution.js";
import { toDisplayString } from "./utils/value-utils.js";

function createBuiltinTools(options: BuildToolOptions, context: ToolInvocationContext): StructuredToolInterface[] {
  return [
    tool(async () => toDisplayString({ now: new Date().toISOString() }), {
      name: "get_time",
      description: "Get current ISO datetime in UTC.",
      schema: z.object({})
    }),
    tool(async (input: { text: string }) => toDisplayString({ echoed: input.text }), {
      name: "echo_text",
      description: "Echo input text for testing tool-calling behavior.",
      schema: z.object({ text: z.string().describe("Text to echo") })
    }),
    tool(async (input: { expression: string }) => {
      const safe = /^[0-9+\-*/().\s]+$/.test(input.expression);
      if (!safe) throw new Error("Expression contains unsupported characters");
      const result = Function(`"use strict"; return (${input.expression});`)();
      if (typeof result !== "number" || Number.isNaN(result)) throw new Error("Invalid expression result");
      return toDisplayString({ result });
    }, {
      name: "calculate",
      description: "Calculate a simple arithmetic expression using + - * / and parentheses.",
      schema: z.object({ expression: z.string().describe("Expression, e.g. (12+3)*4") })
    }),
    tool(async (input: { content: string; category?: string; confidence?: number }) => {
      if (!options.memoryStore) return "Memory store is not configured.";
      if (!context.threadId) throw new Error("threadId is required for memory operations");
      const fact = await options.memoryStore.createFact(context.threadId, {
        content: input.content,
        category: input.category,
        confidence: input.confidence,
        metadata: context.metadata
      });
      return toDisplayString({
        id: fact.id,
        thread_id: fact.threadId,
        content: fact.content,
        category: fact.category,
        confidence: fact.confidence
      });
    }, {
      name: "remember_fact",
      description: "Persist a fact into thread memory for later turns.",
      schema: z.object({
        content: z.string().describe("Fact to remember for this thread"),
        category: z.string().default("context").describe("Memory fact category"),
        confidence: z.number().min(0).max(1).default(0.7).describe("Confidence score")
      })
    }),
    tool(async () => {
      if (!options.memoryStore) return "Memory store is not configured.";
      if (!context.threadId) throw new Error("threadId is required for memory operations");
      const facts = await options.memoryStore.listFacts(context.threadId, { limit: 20 });
      return toDisplayString(facts.map((fact) => ({
        id: fact.id,
        thread_id: fact.threadId,
        content: fact.content,
        category: fact.category,
        confidence: fact.confidence
      })));
    }, {
      name: "list_memory",
      description: "List saved memory facts for the current thread.",
      schema: z.object({})
    }),
    tool(async () => {
      if (!options.skillRegistry) return "Skill registry is not configured.";
      return toDisplayString(options.skillRegistry.listSkills({
        enabledOnly: Boolean(options.enabledSkills?.length),
        enabledNames: options.enabledSkills
      }));
    }, {
      name: "list_skills",
      description: "List the available skills and their summaries.",
      schema: z.object({})
    }),
    tool(async (input: { name: string }) => {
      if (!options.skillRegistry) return { message: "Skill registry is not configured." };
      const skill = options.skillRegistry.getSkill(input.name);
      if (!skill) throw new Error(`Skill not found: ${input.name}`);
      return skill.content;
    }, {
      name: "read_skill",
      description: "Read the full content of a skill by name.",
      schema: z.object({ name: z.string().describe("Skill name") })
    })
  ];
}

export class DefaultAgentToolRegistry implements AgentToolRegistry {
  private readonly structuredTools = new Map<string, StructuredToolInterface>();
  private readonly localTools: Array<LocalToolSpec> = [];
  private readonly mcpPlugins = new Map<string, McpToolPlugin>();

  registerLocalTool<TSchema extends z.ZodTypeAny>(spec: LocalToolSpec<TSchema>): this {
    this.localTools.push(spec as unknown as LocalToolSpec);
    return this;
  }

  registerStructuredTool(toolDef: StructuredToolInterface): this {
    this.structuredTools.set(toolDef.name, toolDef);
    return this;
  }

  useMcpPlugin(plugin: McpToolPlugin): this {
    this.mcpPlugins.set(plugin.name, plugin);
    return this;
  }

  listMcpPlugins(): McpPluginInfo[] {
    return [...this.mcpPlugins.keys()].sort().map((name) => ({ name }));
  }

  async buildMcpTools(context: McpToolPluginLoadContext = { invocationContext: {} }): Promise<Array<{ plugin: string; tool: StructuredToolInterface }>> {
    const loaded: Array<{ plugin: string; tool: StructuredToolInterface }> = [];
    for (const plugin of this.mcpPlugins.values()) {
      const tools = await plugin.loadTools(context);
      for (const toolDef of tools) {
        loaded.push({ plugin: plugin.name, tool: toolDef });
      }
    }
    return loaded;
  }

  async buildTools(options: BuildToolOptions = {}): Promise<StructuredToolInterface[]> {
    const tools = new Map<string, StructuredToolInterface>();
    const executionCoordinator = new ToolExecutionCoordinator();
    const context: ToolInvocationContext = { threadId: options.threadId, runId: options.runId, metadata: options.metadata };

    for (const [name, existingTool] of this.structuredTools.entries()) {
      tools.set(name, wrapToolWithPolicy(existingTool, options, executionCoordinator));
    }

    for (const builtin of createBuiltinTools(options, context)) {
      tools.set(builtin.name, wrapToolWithPolicy(builtin, options, executionCoordinator));
    }

    for (const local of this.localTools) {
      const dynamicTool = tool(
        async (input) => {
          const result = await local.invoke(input, context);
          return toDisplayString(result);
        },
        {
          name: local.name,
          description: local.description,
          schema: local.schema
        }
      );
      tools.set(dynamicTool.name, wrapToolWithPolicy(dynamicTool, options, executionCoordinator, local));
    }

    const mcpTools = await this.buildMcpTools({
      invocationContext: context,
      services: options.mcpServices
    });
    for (const loaded of mcpTools) {
      tools.set(loaded.tool.name, wrapToolWithPolicy(loaded.tool, options, executionCoordinator));
    }

    const allTools = [...tools.values()];
    if (!options.toolAllowlist || options.toolAllowlist.length === 0) {
      return allTools;
    }
    const allow = new Set(options.toolAllowlist.map((item) => item.trim()).filter(Boolean));
    return allTools.filter((item) => allow.has(item.name));
  }
}

export function registerBuiltinTools(registry: AgentToolRegistry): AgentToolRegistry {
  return registry;
}
